### ISSUE TITLE
MPI compilation fails

### ISSUE THREAD

**Issue by david-guyon** [Caltech-IPAC/Montage#25]
In https://github.com/Caltech-IPAC/Montage/blob/master/lib/src/wcssubs3.9.0_montage/Makefile, the compilation of fileutil.o is not included. It causes the compilation of the MPI version to fail because there is an undefined reference to `getfilesize` which is defined in fileutil.c. After including the compilation of this file everything is compiled with success. 

Why is this file not compiled by default?

**Comment by dokeeffe** [comment]
I'm seeing a problem too building mpi modules 
```make[1]: Entering directory '/montage/Montage'
cp mAdd mAddCube mAddExec mBackground mBgExec mBgModel mConvert mCoverageCheck mDiff mDiffExec mDiffFitExec mFitExec mFitplane mFlattenExec mGetHdr mHdrCheck mHdrtbl mImgtbl mMakeHdr mMakeImg mOverlaps mProject mProjectCube mProjectPP mProjExec mPutHdr mShrink mShrinkCube mSubCube mSubimage mSubset mTANHdr mTblSort mTileHdr mTileImage mProjExecMPI mFitExecMPI mDiffExecMPI mBgExecMPI mAddMPI mAddExecMPI ../bin
cp: cannot stat 'mAddMPI': No such file or directory
cp: cannot stat 'mAddExecMPI': No such file or directory
Makefile:250: recipe for target 'install' failed
```

**Comment by dokeeffe** [comment]
Actually your fix works. I opened a PR

**Comment by dgetux** [comment]
Thanks :+1:

**Comment by david-guyon** [comment]
I realize that this bug has been fixed in January 2018. Thanks. I close this issue.

**Comment by david-guyon** [comment]
I just tried to compile the MPI version and compilation fails. I uncommented lines 62 and 63 in [Makefile.LINUX](https://github.com/Caltech-IPAC/Montage/blob/master/Montage/Makefile.LINUX) to enable the MPI compilation. I got the following error:

```
mBgExecMPI.o: In function `main':
/root/Montage/Montage/mBgExec.c:255: warning: the use of `mktemp' is dangerous, better use `mkstemp' or `mkdtemp'
/root/Montage/Montage/mBgExec.c:354: undefined reference to `tsave'
/root/Montage/Montage/mBgExec.c:405: undefined reference to `tsave'
mBgExecMPI.o: In function `nextImg':
/root/Montage/Montage/mBgExec.c:611: undefined reference to `trestore'
mBgExecMPI.o: In function `nextCorr':
/root/Montage/Montage/mBgExec.c:636: undefined reference to `trestore'
collect2: error: ld returned 1 exit status
Makefile:105: recipe for target 'mBgExecMPI' failed
make[1]: *** [mBgExecMPI] Error 1
```

I did not investigate more on this problem.

**Comment by david-guyon** [comment]
I just tried to compile the MPI version and compilation fails. I uncommented lines 62 and 63 in [Makefile.LINUX](https://github.com/Caltech-IPAC/Montage/blob/master/Montage/Makefile.LINUX) to enable the MPI compilation. I got the following error:

```
mBgExecMPI.o: In function `main':
/root/Montage/Montage/mBgExec.c:255: warning: the use of `mktemp' is dangerous, better use `mkstemp' or `mkdtemp'
/root/Montage/Montage/mBgExec.c:354: undefined reference to `tsave'
/root/Montage/Montage/mBgExec.c:405: undefined reference to `tsave'
mBgExecMPI.o: In function `nextImg':
/root/Montage/Montage/mBgExec.c:611: undefined reference to `trestore'
mBgExecMPI.o: In function `nextCorr':
/root/Montage/Montage/mBgExec.c:636: undefined reference to `trestore'
collect2: error: ld returned 1 exit status
Makefile:105: recipe for target 'mBgExecMPI' failed
make[1]: *** [mBgExecMPI] Error 1
```

I did not investigate more on this problem.

**Comment by david-guyon** [comment]
Line 5 in [Makefile](https://github.com/Caltech-IPAC/Montage/blob/master/lib/src/wcssubs3.9.0_montage/Makefile) contains the inclusion of `fileutil.o`. Therefore it has been merged/fixed. Now the compilation error seems to have a different origin.

**Comment by dokeeffe** [comment]
Sorry David, I'm not an expert in this area at all. I was compiling with MPI flags because it was a requirement for lemon https://github.com/vterron/lemon

**Comment by fardal** [comment]
I also encountered the error @david-guyon reported two posts above.  My workaround was to remove all references to mBgExecMPI in Montage/Makefile.Darwin (I'm on a Mac, use the makefile matching your OS).  The other MPI programs compiled correctly.

I note also the documentation instructions at http://montage.ipac.caltech.edu/docs/gridtools.html are incorrect about how to get MPI to compile - if you just change Makefile itself, it will immediately get overwritten by your OS-specific version such as Makefile.Darwin.

**Comment by mroldrobot** [comment]
I got the same error as described by @david-guyon and @fardal .  The workaround mentioned by @fardal works for me as well, but would be better to compile everything correctly, including mBgExecMPI .

**Comment by JohnGood** [comment]
While we have maintained a presence for Montage on GitHub for a several years now,
until recently the actual source code control and issue tracking were done through
other, institution-specific, tools. So while we addressed all issues in a timely
manner, we were regrettably lax about cleaning up these issues in GitHub.

The work done on MPI was by one of our collaborators at JPL.  None of the rest of us
have any experience with MPI or even an MPI system to test on.   In addition, the 
typical Montage workflow is embarrassingly parallel at the process level and there are
several workflow managers that have implemented flows around it.

We believe this is the best way forward for general use as it fits with pretty much every
high-performance computing environment around.  

If there is someone who would like to take on the task of properly integrating MPI into
the Montage modules or add modules (probably high-end executive processes), we 
would be happy to collaborate in an advisory capacity.  However, for now we are 
closing this issue.
